docs/user-guide/setting-up-ci-cd.rst

```diff
@@ -0,0 +1,212 @@
+:desc: Set up a CI/CD pipeline to ensure that iterative improvements are tested and deployed with minimum manual effort
+
+.. _setting-up-ci-cd:
+
+Setting up CI/CD
+================
+
+Developing a bot looks a little different than developing a traditional
+software application, but that doesn't mean you should abandon software
+development best practices. Setting up CI/CD ensures that incremental updates
+to your bot are improving it, not harming it.
+
+.. contents::
+   :local:
+   :depth: 2
+
+
+Overview
+--------
+
+Continous Integration (CI) is the practice of merging in code changes
+frequently, and automatically testing changes as they are committed. Continuous
+Deployment (CD) means automatically deploying integrated changes to a staging
+or production environment. Together, they allow for more frequent improvements
+to your assistant with less manual effort expended on testing and deployment.
+
+This guide will cover **what** should go in a CI/CD pipeline, specific to a
+Rasa project. **How** you implement that pipeline is up to you. There are many
+CI/CD tools out there, and you might already have a preferred one. Many Git repository 
+hosting services like Github, Gitlab, and Bitbucket provide
+their own CI/CD tools. It is recommended to incorporate
+your CI/CD pipeline into a Git repository set up, instead of using a seperate service.
+
+
+
+Continuous Integration
+----------------------
+
+Assistants are best improved with frequent, `incremental updates
+<https://rasa.com/docs/rasa-x/user-guide/improve-assistant/>`_.
+No matter how small a change is, you want to be sure that it doesn't introduce
+new problems or negatively impact the performance of your assistant. Most tests are
+quick enough to run on every change. However, you can set some more
+resource-intentsive tests to run only when certain files have been changed or
+when a certain tag is present.
+
+CI checks usually run on commit, or on merge/pull request.
+
+.. contents::
+   :local:
+
+Validate Data
+#############
+
+:ref:`Data validation <validate-files>` verifies that there are no mistakes or
+major inconsistencies in your domain file, NLU data, or story data. 
+
+.. code-block:: bash
+
+   rasa data validate --fail-on-warnings
+
+If ``rasa
+data validate`` results in errors, training a model will also fail. By
+including the ``--fail-on-warnings`` flag, the command will also fail on
+warnings about problems that won't prevent training a model, but might indicate
+messy data, such as actions listed in the domain that aren't used in any
+stories.
+
+Validate Stories
+################
+
+:ref:`Story validation <test-story-files-for-conflicts>` checks if you have any
+stories where different bot actions follow from the same dialogue history.
+
+.. code-block:: bash
+
+   rasa data validate --fail-on-warnings --max-history 5
+
+Conflicts between stories will prevent Rasa from learning the correct pattern.
```

Comment:
I think it would be worthwhile to give an example of a conflict like we do here:
https://colab.research.google.com/drive/1We1EcSp_uA31OooC78sReryhMCkOjEKM#scrollTo=t-QaPJrUjxT-

Reply:
There's an example at the linked to section of the docs about story validation - I don't think we need to duplicate it

Reply:
That works!